### PR TITLE
Experimental loading of MLX files

### LINKED ISSUE #29333: Add support for models trained using MLX
### Feature request

It would be great if model weights trained and saved through MLX could be easily loaded using the Transformers library. This way, MLX users and Transformers users can more freely collaborate when making open source models.

### Motivation

Currently, Transformers only supports safetensors files with metadata for format in `["pt", "tf", "flax"]`. If Transformers can add `"mlx"` to the list, and if MLX can add `metadata={"format":"mlx"}` when saving safetensors, this can be achieved.

Here is a related issue I created on the `mlx_lm` side:
https://github.com/ml-explore/mlx/issues/743#issuecomment-1965427589

Here is a related PR on the `mlx_lm` repo: https://github.com/ml-explore/mlx-examples/pull/496

### Your contribution

I can send a PR to modify the exact code shortly after.

### PR DESCRIPTION
**Pull request by pcuenca**
# What does this PR do?

Meant as a working draft to support the discussion in #29333.
Fixes #29333 

This code works for a simple roundtrip conversion in `float16`:
```
python -m mlx_lm.convert --hf-path TinyLlama/TinyLlama-1.1B-Chat-v1.0 --dtype float16 --mlx-path tiny-llama-chat
```

However:
- It requires `mlx` to be installed because it's detected as a distinct framework. We should possibly override with `pt` in this case, as it doesn't make sense to demand `mlx` as a dependency. This may or may not clash with recent work done in #29406.
- Many model weights in MLX adopt the same naming conventions as `transformers`, this was a decision made in the `mlx-examples` project to increase interoperability. However, this is not guaranteed. We should probably error if we find incompatible weights.
- I don't know how (or if) this works with quantized weights. Many MLX repos are about creating 4-bit quantized versions that run efficiently on Apple silicon.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@alexweberk, @ArthurZucker

cc @Vaibhavs10

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3328,6 +3328,9 @@ def from_pretrained(
             elif metadata.get("format") == "flax":
                 from_flax = True
                 logger.info("A Flax safetensors file is being loaded in a PyTorch model.")
+            elif metadata.get("format") == "mlx":
+                # This is a mlx file, we assume weights are compatible with pt
+                pass
```
Comment (pcuenca) [inline src/transformers/modeling_utils.py:3333]: This works, but it goes down a code path that requires `mlx` to be installed via https://github.com/huggingface/transformers/blob/63caa370e6c618dbe7d3fd4cbf545cc32eca1a15/src/transformers/utils/generic.py#L118.

I think a better solution in this case could be to replace the `format` with `pt`.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -505,7 +505,7 @@ def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
         # Check format of the archive
         with safe_open(checkpoint_file, framework="pt") as f:
             metadata = f.metadata()
-        if metadata.get("format") not in ["pt", "tf", "flax"]:
+        if metadata.get("format") not in ["pt", "tf", "flax", "mlx"]:
```
Comment (pcuenca) [inline src/transformers/modeling_utils.py:508]: This wouldn't be necessary if we replace the `"format"` as discussed in another comment.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (pcuenca) [reply src/transformers/modeling_utils.py:3302]: I was mistaken, `mlx` is only used if installed, otherwise it works fine.

**Comment by pcuenca** [comment]
This is now ready for review. It essentially fixes `from_pretrained` to allow loading models saved with `mlx`, and adds a simple test. Quantized models can't be loaded, we can explore it further in another PR.

File: tests/test_modeling_utils.py [hunk 3]
```diff
@@ -1256,6 +1256,26 @@ def test_modifying_model_config_causes_warning_saving_generation_config(self):
             self.assertEqual(len(logs.output), 1)
             self.assertIn("Your generation config was originally created from the model config", logs.output[0])
 
+    @require_safetensors
+    def test_model_from_pretrained_from_mlx(self):
+        model = AutoModelForCausalLM.from_pretrained("pcuenq/tiny-random-mistral-mlx")
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply tests/test_modeling_utils.py:1263]: Yes, awesome!

**Review by LysandreJik — APPROVED** [review]
LGTM

**Comment by pcuenca** [comment]
Gentle ping @amyeroberts or @LysandreJik for merge or additional feedback.

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for adding this capability!

**Pull request merged**